src/protocol/sort/multi_bit_permutation.rs

```diff
@@ -91,55 +81,120 @@ pub async fn multi_bit_permutation<'a, F: Field, S: SecretSharing<F>, C: Context
     Ok(one_off_permutation)
 }
 
-/// For a given `idx` check if each of the record has same value as idx.
-/// Steps
-/// 1. Get bit representation of `idx`
-/// 2. Calculate equality check
-///   i. keep record bit if `idx` bit is 1
-///   ii.toggle record bit if idx bit is 0 (Done by taking `share_of_one` - value)
-/// 3. Multiply equality checks for all bits per record - in clear per record this will be 1 only for 1 index while 0 for all others
-async fn get_bit_equality_checkers<F, C, S>(
-    idx: usize,
-    multi_bit_input: &[Vec<S>],
+///
+/// This function accepts a sequence of N secret-shared bits.
+/// When considered as a bitwise representation of an N-bit unsigned number, it's clear that there are exactly
+/// `2^N` possible values this could have.
+/// This function checks all of these possible values, and returns a vector of secret-shared results.
+/// Only one result will be a secret-sharing of one, all of the others will be secret-sharings of zero.
+async fn check_everything<F, C, S>(
     ctx: C,
+    record_idx: usize,
+    input: &[Vec<S>],
 ) -> Result<Vec<S>, Error>
 where
     F: Field,
     C: Context<F, Share = S>,
     S: SecretSharing<F>,
 {
-    let num_multi_bits = multi_bit_input.len();
-    assert_ne!(num_multi_bits, 0);
-    let num_records = multi_bit_input[0].len();
+    let num_bits = input.len();
 
-    let ctx_across_bits = ctx.narrow(&MultiplyAcrossBits);
-    let idx_in_bits = &idx.view_bits::<Lsb0>()[..num_multi_bits];
+    let precomputed_combinations =
+        pregenerate_all_combinations(ctx, record_idx, input, num_bits).await?;
 
-    try_join_all(
-        (0..num_records)
-            .zip(repeat(ctx_across_bits))
-            .map(|(rec, ctx)| async move {
-                let share_of_one = ctx.share_of_one();
-
-                let mult_input = zip(multi_bit_input, idx_in_bits)
-                    .map(|(single_bit_input, bit)| {
-                        if *bit {
-                            single_bit_input[rec].clone()
-                        } else {
-                            -single_bit_input[rec].clone() + &share_of_one
-                        }
-                    })
-                    .collect::<Vec<_>>();
-
-                multiply_all_shares(
-                    ctx,
-                    RecordId::from(idx * num_records + rec),
-                    mult_input.as_slice(),
-                )
-                .await
-            }),
-    )
-    .await
+    // This loop just iterates over all the possible values this N-bit input could potentially represent
+    // and checks if the bits are equal to this value. It does so my computing a linear combination of the
+    // pre-computed coefficients.
+    //
+    // Observe that the coefficients necessary to compute a given equality check form a Sierpiński triangle
+    // <https://en.wikipedia.org/wiki/Sierpi%C5%84ski_triangle#/media/File:Multigrade_operator_AND.svg>
```

Comment:
// Observe that whether a given precomputed coefficient contributes to a
// given equality check follows a Sierpiński triangle
// <https://en.wikipedia.org/wiki/Sierpi%C5%84ski_triangle#/media/File:Multigrade_operator_AND.svg>.
//
// For example, for a three bit value, we have the following:
// ```
// 1 | 1 | 1 | 1 | 1 | 1 | 1 | 1
// 0 | 1 | 0 | 1 | 0 | 1 | 0 | 1
// 0 | 0 | 1 | 1 | 0 | 0 | 1 | 1
// 0 | 0 | 0 | 1 | 0 | 0 | 0 | 1
// 0 | 0 | 0 | 0 | 1 | 1 | 1 | 1
// 0 | 0 | 0 | 0 | 0 | 1 | 0 | 1
// 0 | 0 | 0 | 0 | 0 | 0 | 1 | 1
// 0 | 0 | 0 | 0 | 0 | 0 | 0 | 1
// ```
// This can be computed from row (i) and column (j) indices with `i & j == i`
//
// The sign of the inclusion is less obvious, but we discovered that this
// can be found by taking the same row (i) and column (j) indices:
// 1. Invert the row index and bitwise AND the values: `a = !i & j`
// 2. Count the number of bits that are set: `b = a.count_ones()`
// 3. An odd number means a positive coefficient; an odd number means a negative.
//
// For example, for a three bit value, step 1 produces (in binary):
// ```
// 000 | 001 | 010 | 011 | 100 | 101 | 110 | 111
// 000 | 000 | 010 | 010 | 100 | 100 | 110 | 110
// 000 | 001 | 000 | 001 | 100 | 101 | 100 | 101
// 000 | 000 | 000 | 000 | 100 | 100 | 100 | 100
// 000 | 001 | 010 | 011 | 000 | 001 | 010 | 011
// 000 | 000 | 010 | 010 | 000 | 000 | 010 | 010
// 000 | 001 | 000 | 001 | 000 | 001 | 000 | 001
// 000 | 000 | 000 | 000 | 000 | 000 | 000 | 000
// ```
// Where 000, 101, 011, and 110 mean positive contributions, and
// 001, 010, 100, and 111 mean negative contributions.